signer/src/network/libp2p/swarm.rs

```diff
@@ -0,0 +1,215 @@
+use std::hash::{DefaultHasher, Hash, Hasher};
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::context::Context;
+use crate::keys::PrivateKey;
+use libp2p::autonat::v2::client::{
+    Behaviour as AutoNatClientBehavior, Config as AutoNatClientConfig,
+};
+use libp2p::autonat::v2::server::Behaviour as AutoNatServerBehavior;
+use libp2p::identity::Keypair;
+use libp2p::kad::store::MemoryStore;
+use libp2p::swarm::NetworkBehaviour;
+use libp2p::{
+    gossipsub, identify, kad, mdns, noise, ping, relay, tcp, yamux, Multiaddr, PeerId, Swarm,
+    SwarmBuilder,
+};
+use rand::rngs::OsRng;
+use tokio::sync::Mutex;
+
+use super::errors::SignerSwarmError;
+use super::event_loop;
+
+/// Define the behaviors of the [`SignerSwarm`] libp2p network.
+#[derive(NetworkBehaviour)]
+pub struct SignerBehavior {
+    pub gossipsub: gossipsub::Behaviour,
+    mdns: mdns::tokio::Behaviour,
+    kademlia: kad::Behaviour<MemoryStore>,
```

Comment:
What do we need to store in kademlia?

Reply:
We don't specifically store anything in Kademlia, but the Kademlia protocol includes efficient peer discovery and propogation, so it's generally the recommended peer discovery mechanism. mDNS can only discover local nodes, and rendezvouz is "centralized".

Reply:
Oh interesting. So just adding the kademlia behavior here gives us peer discover that way. Nice, thanks for the info!